File: packages/config/src/env.ts
import { z } from "zod";

const envSchema = z.object({
  API_HOST: z.string().default("0.0.0.0"),
  API_PORT: z.coerce.number().default(4000),
  APP_URL: z.string().url(),
  BETTER_AUTH_SECRET: z.string().min(16),
  BETTER_AUTH_URL: z.string().url(),
  BIGRAG_URL: z.string().url().default("http://localhost:6000"),
  DATABASE_URL: z.string().url(),
  ENCRYPTION_SECRET: z.string().min(32),
  ENCRYPTION_SECRET_PREVIOUS: z.string().min(32).optional(),
  NEXT_PUBLIC_API_URL: z.string().url(),
  NODE_ENV: z
    .enum(["development", "production", "test"])
    .default("development"),
  REDIS_URL: z.string().url()
});

export type Env = z.infer<typeof envSchema>;

export const parseEnv = (
  source: Record<string, string | undefined> = process.env
): Env => {
  const result = envSchema.safeParse(source);
  if (!result.success) {
    const formatted = result.error.flatten().fieldErrors;
    const missing = Object.entries(formatted)
      .map(([key, errors]) => `  ${key}: ${errors?.join(", ")}`)
      .join("\n");
    throw new Error(`Invalid environment variables:\n${missing}`);
  }
  return result.data;
};

File: internal/config/env.go
package config

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Env struct {
	APIHost                    string
	APIPort                    int
	AppURL                     string
	BetterAuthSecret           string
	BetterAuthURL              string
	DatabaseURL                string
	EncryptionSecret           string
	EncryptionSecretPrevious   string
	NextPublicAPIURL           string
	NodeEnv                    string
	RedisURL                   string
}

func ParseEnv() (*Env, error) {
	env := &Env{
		APIHost: getEnvDefault("API_HOST", "0.0.0.0"),
		APIPort: getEnvInt("API_PORT", 4000),
		AppURL:  os.Getenv("APP_URL"),
		BetterAuthSecret: os.Getenv("BETTER_AUTH_SECRET"),
		BetterAuthURL:    os.Getenv("BETTER_AUTH_URL"),
		DatabaseURL:      os.Getenv("DATABASE_URL"),
		EncryptionSecret: os.Getenv("ENCRYPTION_SECRET"),
		EncryptionSecretPrevious: os.Getenv("ENCRYPTION_SECRET_PREVIOUS"),
		NextPublicAPIURL: os.Getenv("NEXT_PUBLIC_API_URL"),
		NodeEnv:          getEnvDefault("NODE_ENV", "development"),
		RedisURL:         os.Getenv("REDIS_URL"),
	}

	var missing []string
	if env.AppURL == "" {
		missing = append(missing, "APP_URL")
	}
	if env.BetterAuthSecret == "" || len(env.BetterAuthSecret) < 16 {
		missing = append(missing, "BETTER_AUTH_SECRET (min 16 chars)")
	}
	if env.BetterAuthURL == "" {
		missing = append(missing, "BETTER_AUTH_URL")
	}
	if env.DatabaseURL == "" {
		missing = append(missing, "DATABASE_URL")
	}
	if env.EncryptionSecret == "" || len(env.EncryptionSecret) < 32 {
		missing = append(missing, "ENCRYPTION_SECRET (min 32 chars)")
	}
	if env.NextPublicAPIURL == "" {
		missing = append(missing, "NEXT_PUBLIC_API_URL")
	}
	if env.RedisURL == "" {
		missing = append(missing, "REDIS_URL")
	}

	if len(missing) > 0 {
		return nil, fmt.Errorf("invalid environment variables:\n  %s", strings.Join(missing, "\n  "))
	}

	return env, nil
}

func (e *Env) IsProduction() bool {
	return e.NodeEnv == "production"
}

func getEnvDefault(key, fallback string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return fallback
}

func getEnvInt(key string, fallback int) int {
	v := os.Getenv(key)
	if v == "" {
		return fallback
	}
	n, err := strconv.Atoi(v)
	if err != nil {
		return fallback
	}
	return n
}
